engine/access/rpc/backend/connection_factory.go

```diff
@@ -226,16 +260,26 @@ func (cf *ConnectionFactoryImpl) invalidateAPIClient(address string, port uint)
 	}
 }
 
+// Close closes the CachedClient connection
 func (s *CachedClient) Close() {
 	s.mutex.Lock()
 	conn := s.ClientConn
 	s.ClientConn = nil
 	s.mutex.Unlock()
+
 	if conn == nil {
 		return
 	}
-	// allow time for any existing requests to finish before closing the connection
-	time.Sleep(s.timeout + 1*time.Second)
+
+	// Mark the connection for closure
+	s.closeRequested.Store(true)
+
+	// If there are ongoing requests, wait for them to complete
+	if s.requestCounter.Load() > 0 {
+		<-s.done
```

Comment:
I'm a little concerned with the blocking call to `Close()`. I think it's correct to implement it this way, but the existing code that calls `Close()` needs to handle it properly. I don't think it does today. I think there are 2 places:
In `invalidateAPIClient`:
https://github.com/onflow/flow-go/blob/2528190b0143af7ca5680975b3a2cd945bd5db86/engine/access/rpc/backend/connection_factory.go#L218-L227

and in the evict handler:
https://github.com/onflow/flow-go/blob/2528190b0143af7ca5680975b3a2cd945bd5db86/engine/access/rpc/engine.go#L151-L158

In the invalidate case, this would cause the request to block, potentially causing it to timeout before retrying on other backends. In the evict handler case, this would cause the add operation called when setting up the new connection to block, similarly delaying the client request. As long as the client object is removed from the cache first, I think it's ok to do the close async.